src/ty/fields.rs

```diff
@@ -78,22 +78,22 @@ pub struct Field<T: Form = MetaForm> {
         feature = "serde",
         serde(skip_serializing_if = "Option::is_none", default)
     )]
-    name: Option<T::String>,
+    pub name: Option<T::String>,
     /// The type of the field.
     #[cfg_attr(feature = "serde", serde(rename = "type"))]
-    ty: T::Type,
+    pub ty: T::Type,
     /// The name of the type of the field as it appears in the source code.
     #[cfg_attr(
         feature = "serde",
         serde(skip_serializing_if = "Option::is_none", default)
     )]
-    type_name: Option<T::String>,
+    pub type_name: Option<T::String>,
     /// Documentation
     #[cfg_attr(
         feature = "serde",
         serde(skip_serializing_if = "Vec::is_empty", default)
     )]
-    docs: Vec<T::String>,
+    pub docs: Vec<T::String>,
```

Comment:
These being `pub` makes the impl's (`name()`. `type_name()` etc) sortof pointless. We'd have to decide here whether we are happy to expose this internal structure. Maybe `pub(crate)` for now if it's useful to avoid allocating when retaining, but it looks like you just `clone()` each type anyway, so perthaps these aren't needed at all? Or perhaps functions like `id_mut()` etc to go along with the existing non-mut ones is a way to go?

Reply:
Or maybe the way to go is to just clone each field and construct a new thing with `::new()` in each case? Not sure; I can see a loaf od stuff has been made `pub` in the file, and I thnk to avoid exposing a lot of internal details maybe just `pub(crate)` for now is easiest? (and you could update each type in-place and `mem::swap` or something to avoid allocations, maintaining a set of types you've visited so you never look at a type more than once anyway)

Reply:
I've picked `pub(crate)` for the path of least resistance without needing to add `mut` counterparts for the getter functions or remove them entirely. Also, I've implemented a dummy `Type::Default` which assumes the default type is a `bool` to avoid extra allocations, thanks!

Reply:
The other option of course would have be to implement methods directly on the types, in a similar manner to the `InfoPortable` trait implementations. OTOH it may be a good thing to keep those implementations inside `retain`.

Taking the `IntoPortable` comparison a bit further, imagine a trait `RetainTypes`, implemented in a similar way, and then you would have to manually implement it in the same way for the metadata definitions. Then you could have an API like `PortableRegistry::retain_for<T: RetainTypes>(metadata_type: &T)`, obviously would still need to be able to provide multiple types somehow, but I think you get the drift.